tests/cases/compiler/reducibleIndexedAccessTypes.ts

```diff
@@ -0,0 +1,43 @@
+// @strict: true
+// @noEmit: true
+
+// Repro from 53030
+
+enum Type { A, B, C }
+
+interface PayloadStructure {
+    dataType: Type;
+    data: unknown;
+}
+
+interface PayloadA extends PayloadStructure {
+    dataType: Type.A;
+    data: string;
+}
+
+interface PayloadB extends PayloadStructure {
+    dataType: Type.B;
+    data: number;
+}
+
+interface PayloadC extends PayloadStructure {
+    dataType: Type.C;
+    data: {
+        x: number;
+        y: number;
+    };
+}
+
+type Payload = PayloadA | PayloadB | PayloadC
+
+type MappedPayload2 = {
+    [K in Type]?: (data: (Payload & { dataType: K })["data"]) => void
+}
+
+const payloads2: MappedPayload2 = {
+    [Type.A]: data => { console.log(data) }
+}
+
+// Additional repro from 53030
+
+type GetPayload<P extends Payload, K extends keyof P> = P extends { dataType: K } ? P["data"] : never;

```

Comment:
perhaps it would still be worthwhile to include a test from #51161? that one was even more suspicious as the index access in the true branch was using the same property name as the one in the `extendsType`. This additional repro is using a different one

```suggestion
type GetPayload<P extends Payload, K extends keyof P> = P extends { dataType: K } ? P["data"] : never;

// repro from 51161

export type AnyOneof = { oneofKind: string; [k: string]: unknown } | { oneofKind: undefined };
export type AnyOneofKind<T extends AnyOneof> = T extends { oneofKind: keyof T }
    ? T['oneofKind']
    : never;

```

Reply:
Sure, I'll add the extra test.